README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# My modified OncoBayes2

<!-- badges: start -->
<!-- badges: end -->

This is my modified version of the [original
OncoBayes2](https://CRAN.R-project.org/package=OncoBayes2) package.

The most significant difference is the support of providing correlation
parameter when sampling the component via the newly added parameter
`prior_EX_corr_mu_comp` in `blrm_exnex()`.

In the original package,
$\theta = (log \alpha, log \beta) \sim N(\mu, \Sigma)$ where the
**covariance** matrix $\Sigma$ is a diagonal matrix whose diagonals are
provided via `prior_EX_mu_sd_comp`(Note, this parameter provides the
**standard variance**).

Now, the off-diagonal part of $\Sigma$ can also be specified by
supplying the **correlation** via `prior_EX_corr_mu_comp`.

## NOTE

1.  This is just a **proof of concept** during my replication of Table
    14-7 in this [trial
    protocol](https://clinicaltrials.gov/ProvidedDocs/64/NCT02108964/Prot_001.pdf).
    The original OncoBayes2 misses the functionality to accept user
    defined correlation, which is also reported on the
    [internet](https://stats.stackexchange.com/questions/617633/correlation-for-priors-in-blrm-with-ewoc-in-r-package-oncobayes2/626132#626132)

2.  Currently this package is forked and modified from [CRAN’s github
    version of OncoBayes2](https://github.com/cran/OncoBayes2). After
    installation it will **overwrite** your existing OncoBayes2.

3.  This is currently WIP. Many aspect, like the `prior_summary` method
    should be updated accordingly to reflect the correlation
    information. Also many document template is missing hence the manual
    cannot be auto-updated properly.

## Installation

You can install the development version of OncoBayes2 like so:

``` r
remotes::install_github("fenguoerbian/MyOncoBayes2")
```

## Example

This is a basic example which reproduce the results in Table 14-7 in
this [trial
protocol](https://clinicaltrials.gov/ProvidedDocs/64/NCT02108964/Prot_001.pdf).

``` r
library(OncoBayes2)
#> This is OncoBayes2 version 0.8.123456 (released 2023-07-21, git-sha e5a8128)
## basic example code
```

### Setup MCMC options

``` r
set.seed(123)
.user_mc_options <- options(OncoBayes2.MC.warmup=500, OncoBayes2.MC.iter=20000, OncoBayes2.MC.chains=10,
                            OncoBayes2.MC.save_warmup=FALSE, mc.cores = 10)
```

### Setup basic trial infomation

``` r
SA_trial_setup <- blrm_trial(
    data = tibble::tibble(group_id = as.factor("All"),
                  drug_A = 50,
                  num_patients = NA,
                  num_toxicities = NA),
    # dose-toxicity data available at design stage of trial.
    
    dose_info = tibble::tibble(group_id = as.factor("All"), 
                       drug_A = c(50,75,150,300,450,600,800,1000), 
                       dose_id = c(1, 2, 3, 4, 5, 6, 7, 8), stratum_id = "all"), 
    # specification of the dose levels as planned for the ongoing trial arms.
    
    drug_info = tibble::tibble(drug_name = "drug_A", 
                       dose_ref = 300, 
                       dose_unit = "ug/kg"
                       ,
                       reference_p_dlt = exp(-3.068) / (1 + exp(-3.068))
    )
    # specification of drugs used in trial arms
    
    ,simplified_prior = FALSE
    ,EXNEX_comp=TRUE
    ,EXNEX_inter=FALSE
    ,interval_prob = c(0,0.16,0.33,1)
    ,interval_max_mass = c(prob_underdose = 1 # The prob of under-dose is allowed maximum to 1.
                           , prob_target = 1 # The prob of target-dose is allowed maximum to 1.
                           , prob_overdose = 0.25) # The prob of over-dose is allowed maximum to 0.28.
)
#> Please configure blrm_exnex using the update() function.


dims <- summary(SA_trial_setup,"dimensionality")
num_comp <- dims$num_components
```

### Update the prior

``` r
dims <- summary(SA_trial_setup,"dimensionality")
num_comp <- dims$num_components

SA_trial_start <- update(
    SA_trial_setup,
    ##component1 MAP prior
    prior_EX_mu_mean_comp = matrix(
        c(-3.068, # mean of intercept 
          0.564), # mean of log-slope 
        nrow = num_comp,
        ncol = 2
    ),
    prior_EX_mu_sd_comp = matrix(
        c(2.706, # sd of intercept
          0.728), # sd of log-slope
        nrow = num_comp,
        ncol = 2
    ),
    prior_EX_corr_mu_comp = -0.817,    # HERE! supply the CORRELATION!
    prior_EX_tau_mean_comp = matrix(
        c(0, 0),
        nrow = num_comp,
        ncol = 2
    ),
    prior_EX_tau_sd_comp = matrix(
        c(1, 1),
        nrow = num_comp,
        ncol = 2
    ),
    prior_EX_prob_comp = matrix(1, nrow = 1, ncol = 1),
    prior_tau_dist = 0, # single-agent, without historical data, set this to 0.
                        # see single-agent example in this package for more details
    prior_PD = FALSE
)
```

**NOTE:** `prior_summary` needs to be updated to reflect user supplied
correlation

``` r
prior_summary(SA_trial_start)
#> Bayesian Logistic Regression Model with EXchangeability-NonEXchangeability
#> 
#> Mixture configuration
#> ---------------------
#> EXNEX components : 0 
#> component
#> I(log(drug_A/300)) 
#>                  0 
#> 
#> Prior probability for exchangeability per group
#>      component
#> group I(log(drug_A/300))
#>   All                  1
#> 
#> EXchangable hyperparameter priors
#> ---------------------------------
#> Component parameters
#> Mean mu_log_beta
#>                    coefficient intercept       log_slope      
#>                    prior            mean    sd      mean    sd
#> component                                                     
#> I(log(drug_A/300))                 -3.07  2.71      0.56  0.73
#> 
#> Heterogeneity tau_log_beta (known)
#>                    coefficient intercept    log_slope   
#>                    stratum           all          all   
#>                    prior            mean sd      mean sd
#> component                                               
#> I(log(drug_A/300))                     0  1         0  1
#> 
#> Correlation LKJ
#> component
#> I(log(drug_A/300)) 
#>                  1 
#> 
#> Model has no interaction parameters.
#> 
#> NonEXchangable priors
#> ---------------------
#> Component parameters
#> Mean mu_log_beta
#>                    coefficient intercept       log_slope      
#>                    prior            mean    sd      mean    sd
#> component                                                     
#> I(log(drug_A/300))                 -3.07  2.71      0.56  0.73
#> 
#> Model has no interaction parameters.
```

### Check the results against Table 14-7

``` r
summary(SA_trial_start, "dose_prediction")
#> # A tibble: 8 × 13
#>   group_id drug_A dose_id stratum_id   mean    sd   `2.5%`   `50%` `97.5%`
#>   <fct>     <dbl>   <dbl> <fct>       <dbl> <dbl>    <dbl>   <dbl>   <dbl>
#> 1 All          50       1 all        0.0776 0.190 8.75e-10 0.00183   0.786
#> 2 All          75       2 all        0.0905 0.202 1.61e- 8 0.00372   0.819
#> 3 All         150       3 all        0.121  0.226 2.23e- 6 0.0128    0.868
#> 4 All         300       4 all        0.172  0.254 2.27e- 4 0.0447    0.907
#> 5 All         450       5 all        0.223  0.269 2.02e- 3 0.0961    0.926
#> 6 All         600       6 all        0.282  0.279 5.91e- 3 0.172     0.939
#> 7 All         800       7 all        0.366  0.291 1.37e- 2 0.289     0.959
#> 8 All        1000       8 all        0.441  0.298 2.37e- 2 0.400     0.975
#> # ℹ 4 more variables: prob_underdose <dbl>, prob_target <dbl>,
#> #   prob_overdose <dbl>, ewoc_ok <lgl>
```

``` r
summary(SA_trial_start, "ewoc_check")
#> # A tibble: 8 × 9
#>   group_id drug_A dose_id stratum_id prob_overdose_est prob_overdose_stat
#>   <fct>     <dbl>   <dbl> <fct>                  <dbl>              <dbl>
#> 1 All          50       1 all                   0.0336            -609.  
#> 2 All          75       2 all                   0.0525            -379.  
#> 3 All         150       3 all                   0.110             -172.  
#> 4 All         300       4 all                   0.225              -57.4 
#> 5 All         450       5 all                   0.335                2.33
#> 6 All         600       6 all                   0.447               58.5 
#> 7 All         800       7 all                   0.589              149.  
#> 8 All        1000       8 all                   0.696              258.  
#> # ℹ 3 more variables: prob_overdose_mcse <dbl>, prob_overdose_ess <dbl>,
#> #   prob_overdose_rhat <dbl>
```


NEWS.md
# OncoBayes2 0.8-9 - July 20th, 2023

## Bug fixes

* resolve compilation from source issues on some platforms triggered by
  changes in rstantools

# OncoBayes2 0.8-8 - March 3rd, 2023

## Enhancements

* upon package load `OncoBayes2` will now report the date of the release
  and the respective git commit hash used to create the sources of the
  package.

## Bug fixes

* ensure compatibility with posterior 1.4.0 (posterior returns `num()`
  formatted columns, which `OncoBayes2` is not supporting)
* ensure C++17 compatiblity per CRAN (triggers an issue with
  clang 16)

# OncoBayes2 0.8-7 - August 24th, 2022

## Enhancements

* add warning messages when printing `blrmfit` objects for divergent
  transitions or non-convergence of parameters
* add diagnostic extraction functions `nuts_params`, `rhat`, `log_posterior`
  and `neff_ratio` for `blrmfit` objects as defined in `bayesplot`
* speedup `pp_data` by about 25% which is used in all posterior
  summary functions
* add `ewoc_check` summary routine for `blrm_trial` objects. The
  returned summary contains MCMC diagnostics and accuracy estimates of
  the EWOC metric as defined for the trial. This allows to assess the
  MCMC estimation error for EWOC.
* added warning for imprecise estimates of the EWOC metric whenever
  `blrm_trial` objects are printed.

## Bug fixes

* fix broken exclusion of unnecessary random variables whenever
  `save_warmup=FALSE`.

# OncoBayes2 0.8-6 - May 2nd, 2022

* allow cmdstanr as new backend for blrm_exnex. The Stan model file is
  written with the function cmdstanr::write_stan_file to disk such
  that the model binaries can be cached if the user defines the global
  option cmdstanr_write_stan_file_dir. See ?cmdstanr::write_stan_file
  (requires cmdstanr >= 0.5.0)

# OncoBayes2 0.8-5 - February 28th, 2022

* fix issue with example_model not exposing properly objects when
  running with silent=TRUE

# OncoBayes2 0.8-4 - February 18th, 2022

* Switch SBC runs to use clustermq in lieu of batchtools. Also now use
  L'Ecuyer CMG as rng engine during SBC runs.

* add critical_quantile function allowing to calculate critical doses
  which fulfill conditions like EWOC.

# OncoBayes2 0.8-3 - October 7th, 2021

* Make Stan model compile with 2.27+ (lupmf postfix for variables is
  not allowed)

# OncoBayes2 0.8-2 - September 13th, 2021

* Drop unneccessary random variables from the posterior whenever
  save_warmup=FALSE. This decreases the size of the posterior in
  memory by ~60% for the combo2 example.

# OncoBayes2 0.8-1 - September 1st, 2021

* Address CRAN comments

# OncoBayes2 0.8-0 - August 31st, 2021

* Significantly speedup Stan model by dropping normalization of binomial
* The default interaction model for blrm_trial is now saturating
  (blrm_formula_saturating) in logit space, improving behavior for combinations 
  with antagonistic toxicity.
* blrm_trial simplified prior now defaults to ENXEX off, and prior parameter
  defaults have been updated.
* New save_warmup argument to blrm_exnex can be used to disable saving of warmup
  samples, which defaults to the OncoBayes2.MC.save_warmup option, which 
  defaults to TRUE in line with previous versions.
  NOTE: Starting with the next version 0.9.0 the new default will be FALSE!
* Do not crash for summary(trial, "data_prediction") if blrm_trial has NULL or 
  0-row data
* Spurious warning for dose_id NA on blrm_trial instantiation removed
* Spurious warning for dose not prespecified on summary(..., newdata=...)
  removed
* New plot function plot_toxicity_intervals_stacked(), which visualizes model
  posterior or predictive probabilities in discrete intervals, over a continuous
  range of doses
* Automated visual regression testing implemented using the vdiffr package
* Fix in plot_toxicity_curve to avoid overplotting if newdata contains columns
  other than grouping variables and doses (e.g. dose_id)
* Avoid re-parsing model formula whenever newdata is not used in
  respective functions. This prevents that changes to global variables
  (like dref) have any effect whenever the fitted data only is used.

# OncoBayes2 0.7-0 - May 7th, 2021

* Catch data specification errors for wrongly nested groups/strata
* BLRM trials now also print toxicity probability intervals and the EWOC setting
* The blrm_formula_linear and blrm_formula_saturating interaction model formula 
  generators can now be customized to generate specific interactions using the
  specific_interaction_terms argument.
* Add predictive summaries to summary method.
* Switch predictive outputs to be based on averaging sampling density. This avoids
  the need for additional sampling for predictive summaries.
* Add plotting methods for blrmfit and blrm_trial objects.

# OncoBayes2 0.6-5 - May 7th, 2020

* Drop RBesT dependency
* Avoid setting the ggplot2 theme upon package load
* Enhance SBC runs to monitor and report sampler diagnostics

# OncoBayes2 0.6-4 - April 7th, 2020

* Add Operation Qualification script (run-oq.R in inst/extra)
* Add saturating interaction model (blrm_formula_saturating) and associated tests

# OncoBayes2 0.6-3 - March 18th, 2020

* 2nd attempt to work around issues found by clang sanitizer in Stan model.

# OncoBayes2 0.6-2 - March 9th, 2020

* Work around issues found by clang sanitizer in Stan model.

# OncoBayes2 0.6-1 - March 4th, 2020

* Fix issue with upcoming R 4.0 which changes stringsAsFactors default
  to FALSE
* Fix for upcoming tibble 3.0 upgrade which changes conventions of
  growing factors within tibbles

# OncoBayes2 0.6-0 - February 11th, 2020

* Fix issue with model outputs when EXNEX is being used. We recommend
  all users to upgrade to this version. The SBC runs now include
  checks for the per-group estimates to avoid a regression of the bug.
* Correct printing of prior information wrt. to summaries of number of
  strata and groups.

# OncoBayes2 0.5-8 - December 12th, 2019

* Merge data with same group and dosing in blrm_trial before passing

R/summary.R
#' Summarise model results
#'
#' Provides model summaries for \code{\link{blrm_exnex}} and
#' \code{\link{blrm_trial}} analyses.
#'
#' @template args-methods
#' @template args-prob
#' @param interval_prob optional vector of sorted quantiles for which
#'     the interval probabilities are calculated
#' @param predictive logical indicates if the posterior predictive is
#'     being summarized. Defaults to \code{FALSE}.
#' @template args-transform
#' @param ... not used in this function
#'
#' @details
#' The calculated posterior summaries are returned as a
#' \code{data.frame} and contain optional interval probabilites for
#' the specified vector of sorted quantiles. These summaries are
#' calculated on the response scale by default and can be obtained on
#' the link scale when setting \code{transform=FALSE}.
#'
#' When the results are requested for the predictive distribution with
#' \code{predictive=TRUE}, then the link scale refers to the total
#' counts while the transformed scale divides the (predictive) counts
#' by the number of trials such that results are on the 0-1 scale.
#'
#' @return Returns a \code{data.frame} of the key summaries of the
#'     posterior mean, standard deviation, central probability
#'     interval, median and optional interval probabilities. Each row
#'     of the \code{data.frame} corresponds to the respective input
#'     data which is by default the same data set as used for the
#'     \code{\link{blrm_exnex}} analysis or the data specified in the
#'     \code{newdata} argument.
#'
#' @template start-example
#' @examples
#' example_model("single_agent", silent=TRUE)
#'
#' ## obtain underdosing (0-0.16], target dosing (0.16-0.33] and
#' ## overdosing (0.33-1] probabilities
#' summary(blrmfit, interval_prob=c(0,0.16,0.33,1))
#'
#' ## obtain predictive distribution for respective cohorts and
#' ## calculate probability for no event, 1 event or >1 event
#' ## note that this does the calculation for the cohort sizes
#' ## as put into the data-set
#' summary(blrmfit, interval_prob=c(-1,0,1,10), predictive=TRUE)
#'
#' ## to obtain the predictive for a cohort-size of 6 for all patients
#' ## in the data-set one would need to use the newdata argument, e.g.
#' summary(blrmfit, newdata=transform(hist_SA, num_patients=6),
#'                  interval_prob=c(-1,0,1,10), predictive=TRUE)
#'
#' @template stop-example
#'
#' @method summary blrmfit
#' @export
summary.blrmfit <- function(object, newdata, transform=!predictive, prob=0.95, interval_prob, predictive=FALSE, ...) {
    assert_numeric(prob, lower=0, upper=1, finite=TRUE, any.missing=FALSE, min.len=1)

    data <- object$data
    if(!missing(newdata))
        data <- newdata
    rn <- rownames(data)
    nr <- nrow(data)

    probs <- sort(unique(c(0.5, 0.5- prob/2, 0.5+ prob/2)))

    ## use the SAS defintion for quantiles whenever the output are
    ## discrete as for the predictive which is type 2. Type 7 is the
    ## default in R which gives non-integer outputs for integer only
    ## data.
    quantile_type <- ifelse(predictive, 2, 7)

    if(predictive) {
        sizes <- pp_binomial_trials(object, newdata=newdata)

        post_prob <- posterior_linpred(object, transform=TRUE, newdata=newdata)

        ## integrate binomial sampling density per outcome and then
        ## summarize the obtained predictive p(y)

        draws <- nrow(post_prob)
        max_size <- max(sizes)
        pred_lpmf <- matrix(-Inf, nrow=max_size+1, ncol=nr)
        pred_pmf <- matrix(0, nrow=max_size+1, ncol=nr)
        pred_cpmf <- matrix(1.0, nrow=max_size+1, ncol=nr)
        for(r in seq_len(nr)) {
            ## create matrix of outcomes x draws of prob
            outcomes_row <- 0:sizes[r]
            nor <- length(outcomes_row)
            Pr <- matrix(post_prob[,r], nrow=nor, ncol=draws, byrow=TRUE)
            pred_lpmf[seq_len(nor),r] <- apply(dbinom(outcomes_row, sizes[r], prob=Pr, log=TRUE), 1, log_mean_exp)
            ## normalize to sum-to-one per column
            pred_lpmf[seq_len(nor),r] <- pred_lpmf[seq_len(nor),r] - matrixStats::logSumExp(pred_lpmf[seq_len(nor),r])
            pred_pmf[,r] <- exp(pred_lpmf[,r])
            pred_cpmf[seq_len(nor),r] <- cumsum(pred_pmf[seq_len(nor),r])
            ## ensure that largest number is truly 1.0
            pred_cpmf[,r]  <- pmin(pred_cpmf[,r], 1.0)
        }

        outcomes <- 0:max_size
        all_outcomes <- matrix(outcomes, nrow=max_size+1, ncol=nr)

        if(transform) {
            ## transform outcomes to 0-1 scaled response rates
            all_outcomes <- sweep(all_outcomes, 2, sizes, "/")
            if(any(sizes == 0))
                warning("Data rows ", paste(which(sizes == 0), collapse=", "), " contain trials of size 0.")
        }

        m <- colSums(all_outcomes * pred_pmf)
        s <- sqrt(colSums(sweep(all_outcomes, 2, m)^2 * pred_pmf))

        quants <- matrix(NA, nrow=nr, ncol=length(probs))
        for(r in seq_len(nr)) {
            quants[r,]  <- findInterval(probs, pred_cpmf[seq_len(sizes[r]+1),r], left.open=FALSE)
        }
        if(transform)
            quants <- sweep(quants, 1, sizes, "/")

        out_sum <- cbind(data.frame(mean=m, sd=s), quants)

        if(!missing(interval_prob)) {
            if(transform) {
                assert_numeric(interval_prob, lower=-1, upper=1, any.missing=FALSE, sorted=TRUE, finite=TRUE)
            } else {
                assert_integerish(interval_prob, any.missing=FALSE, sorted=TRUE)
            }

            nip <- length(interval_prob)
            out_interval <- matrix(NA, nrow=nr, ncol=nip+1)
            for(r in seq_len(nr)) {
                all_outcomes_r <- all_outcomes[seq_len(sizes[r]+1),r]
                outcome_idx <- findInterval(c(-Inf, interval_prob, Inf), all_outcomes_r, left.open=FALSE)
                cpmf <- c(0,pred_cpmf[,r])
                out_interval[r,] <- diff(cpmf[outcome_idx+1])
            }
            out_interval <- out_interval[,-c(1,nip+1),drop=FALSE]
            colnames(out_interval) <- paste0("(", interval_prob[seq_len(nip-1)], ",", interval_prob[1 + seq_len(nip-1)], "]")
            out_sum <- cbind(out_sum, as.data.frame(out_interval))
        }

    } else {
        post <- posterior_linpred(object, transform=transform, newdata=newdata)
        S <- nrow(post)

        if (ncol(post) > 0) {
          out_sum <- as.data.frame(t(apply(post, 2, function(l) {
            if(all(is.na(l)))
              return(rep(NA, times=2+length(probs)))
            c(mean=mean(l), sd=sd(l), quantile(l, probs=probs, type=quantile_type))
          })))
        } else {
          out_sum <- data.frame(0.0 * matrix(ncol = (2+length(probs)), nrow = 0))
        }

        if(!missing(interval_prob)) {
            if(transform) {
                ## negative values are admissable as we otherwise cannot
                ## include 0
                assert_numeric(interval_prob, lower=-1, upper=1, any.missing=FALSE, sorted=TRUE, finite=TRUE)
            } else {
                assert_numeric(interval_prob, any.missing=FALSE, sorted=TRUE, finite=TRUE)
            }

            if (ncol(post) > 0) {
              out_interval <- as.data.frame(t(apply(post, 2, function(l) table(cut(l, breaks=c(-Inf, interval_prob, Inf)))/S )))
            } else {
              out_interval <- data.frame(0.0 * matrix(ncol = (1+length(interval_prob)), nrow = 0))
              colnames(out_interval) <- levels(cut(numeric(0), breaks=c(-Inf, interval_prob, Inf)))
            }

            out_interval <- out_interval[,-c(1,ncol(out_interval)), drop=FALSE]
            out_sum <- cbind(out_sum, out_interval)
        }
    }

    names(out_sum)[seq_len(2+length(probs))] <- c("mean", "sd", paste0(100 * probs, "%"))

    rownames(out_sum) <- rn
    out_sum
}

#' Summarise trial
#'
#' Provides model summaries for \code{\link{blrm_trial}} analyses.
#' @param object \code{\link{blrm_trial}} object
#' @param summarize one of the following options:
#' \itemize{
#'   \item{}{\code{blrmfit}: summary of the underlying blrmfit object with further arguments ...}
#'   \item{}{\code{blrm_exnex_call}: blrm_exnex call used to create the \code{blrmfit} object}
#'   \item{}{\code{drug_info}: drug_info for the trial, contains drugs, reference doses and units}
#'   \item{}{\code{dose_info}: dose_info that were defined}
#'   \item{}{\code{dose_prediction} prediction for the defined \code{dose_info}}
#'   \item{}{\code{data}: data that were observed}
#'   \item{}{\code{data_prediction}: prediction for the observed data}
#'   \item{}{\code{newdata_prediction}: prediction for data provided with the \code{newdata} argument}
#'   \item{}{\code{dimensionality}: numeric vector with entries \code{"num_components"},
#'     \code{"num_interaction_terms"}, \code{"num_groups"}, \code{"num_strata"}}
#'   \item{}{\code{interval_prob}: interval probabilities reported in the standard outputs}
#'   \item{}{\code{interval_max_mass}: named vector defining for each interval of the
#'     \code{interval_prob} vector a maxmimal admissable
#'     probability mass for a given dose level}
#'   \item{}{\code{ewoc_check}: MCMC diagnostic and precision estimates of ewoc defining metrics for the defined doses in \code{dose_info} (default) or for the doses provided in the \code{newdata} argument. Please refer to the details for reported diagnostics.}
#' }
#' @param ... further arguments for \code{\link{summary.blrmfit}}
#'
#' @details The \code{ewoc_check} summary routine allows to assess the
#'     accuracy and reliability of the ewoc criterion with respect to
#'     MCMC sampling noise. The returned summary provides detailled
#'     MCMC convergence and precision estimates for all criteria
#'     defined by \code{interval_prob} and \code{interval_max_mass}
#'     which contribute to EWOC metric. That is, for each interval
#'     probability with a maximal mass of less than unity the routine
#'     will return these columns:
#' \describe{
#' \item{\code{est}}{the MCMC estimate defining the critical value. For intervals defined by a tail probability this corresponds to the respective critical quantile while for interval probabilites this is equal to the interval probability.}
#' \item{\code{stat}}{centered and standardized test quantity. The estimate is centered by the critical value and scaled by the Monte-Carlo standard error (MCSE) of the estimate. Hence, negative (positive) values correspond to the constraint being (not) fulfilled. The standardization with the MCSE allows to compare the values to standard normal quantiles accordingly.}
#' \item{\code{mcse}}{the Monte-Carlo standard error of the estimate determined with \code{\link[posterior]{mcse_quantile}} (tail probability) or \code{\link[posterior]{mcse_mean}} (interval probability) functions.}
#' \item{\code{ess}}{the Monte-Carlo effective sample size of the estimate determined with \code{\link[posterior]{ess_quantile}} (tail probability) or \code{\link[posterior]{ess_mean}} (interval probability) functions.}
#' \item{\code{rhat}}{the Monte-Carlo non-convergence diagnostic Rhat as determined with the \code{\link[posterior]{rhat} function}.}
#' }
#'
#' For the common case of requiring that 33\% DLT probability is not
#' exceeded by more than 25\% of the posterior probability mass, the
#' estimate column \code{est} contains the 75\% quantile
#' \eqn{q_{75\%}}{q75\%} and the standardized statistic \code{stat} is
#' defined as:
#'
#' \deqn{\mbox{stat} = \frac{q_{75\%} - 33\%}{\mbox{mcse}_{q_{75\%}}}}{stat = (q75\% - 33\%)/mcse_q75\%}
#'
#' The statistic is approximately distributed as a standard normal
#' variate. The \code{ewoc_check} summary can be used to ensure that
#' the MCMC estimation accuracy is sufficient.
#'
#' @template start-example
#' @examples
#' # construct initial blrm_trial object from built-in example datasets
#' combo2_trial_setup <- blrm_trial(
#'   data = hist_combo2,
#'   dose_info = dose_info_combo2,
#'   drug_info = drug_info_combo2,
#'   simplified_prior = TRUE
#' )
#'
#' # extract blrm_call to see setup of the prior as passed to blrm_exnex
#' summary(combo2_trial_setup, "blrm_exnex_call")
#'
#' # extract ewoc precision accuracy
#' ec <- summary(combo2_trial_setup, "ewoc_check")
#'
#' # find any ewoc metrics which are within 95% MCMC error of the threshold
#' # these are counted as "imprecise" when printing blrm_trial objects
#' subset(ec, abs(prob_overdose_stat) < qnorm(0.975))
#'
#' # ensure that the ewoc metric only flags "ok" whenever the MCMC error
#' # is with 95% below the threshold
#' ewoc_ok <- ec$prob_overdose_stat < qnorm(0.025)
#'
#' @template stop-example
#'
#' @method summary blrm_trial
#' @export
summary.blrm_trial <- function(
  object,
  summarize = c(
    "blrmfit",
    "blrm_exnex_call",
    "data",
    "drug_info",
    "dose_info",
    "dose_prediction",
    "data_prediction",
    "newdata_prediction",
    "dimensionality",
    "interval_prob",
    "interval_max_mass",
    "ewoc_check"
  ),
  ...
) {
  args <- list(...)
  if (missing(summarize)) {
    if ("newdata" %in% names(args)) {
      summarize <- "newdata_prediction"
    } else {
      summarize <- "data_prediction"
    }
  } else {
    summarize <- match.arg(summarize)
  }



  if (summarize %in% c("blrmfit", "blrm_exnex_call", "dose_prediction", "data_prediction", "newdata_prediction", "ewoc_check")) {
    .assert_is_blrm_trial_and_prior_is_set(object)
  } else {
    .assert_is_blrm_trial(object)
  }

  newdata_predict <- function() {
    args[["newdata"]] <- .blrm_trial_sanitize_data(
      trial=object,
      data=args[["newdata"]],
      warning_if_dose_not_prespecified = FALSE,
      require_num_patients = FALSE,
      require_num_toxicities = FALSE
    )

    args_without_newdata <- args
    args_without_newdata[["newdata"]] <- NULL
    do.call(".blrm_trial_predict", c(list(object, args[["newdata"]]), args_without_newdata))
  }

  ewoc_check <- function() {
      if(!("newdata" %in% names(args))) {
          .blrm_trial_ewoc_check(object, object$ewoc_check)
          return(object$ewoc_check)
      }

      args[["newdata"]] <- .blrm_trial_sanitize_data(
          trial=object,
          data=args[["newdata"]],
          warning_if_dose_not_prespecified = FALSE,
          require_num_patients = FALSE,
          require_num_toxicities = FALSE
      )

      args_without_newdata <- args
      args_without_newdata[["newdata"]] <- NULL
      ewoc_check <- do.call(".blrm_trial_compute_ewoc_check", c(list(object, newdata=args[["newdata"]]), args_without_newdata))
      .blrm_trial_ewoc_check(object, ewoc_check)
      ewoc_check
  }

  switch(summarize,
    blrmfit = summary(object$blrmfit, ...),
    blrm_exnex_call = object$blrmfit$call,
    drug_info = object$drug_info,
    dose_info = object$dose_info,
    dose_prediction = .blrm_trial_predict(object, object$dose_info, ...),
    data = object$data,
    data_prediction = .blrm_trial_predict(object, object$data, ...),
    newdata_prediction = newdata_predict(),
    dimensionality = object[c("num_components", "num_interaction_terms", "num_groups", "num_strata")],
    interval_prob = object$interval_prob,
    interval_max_mass = object$interval_max_mass,
    ewoc_check = ewoc_check()
  )
}
